From 5e25a49992c2eaa762c062a01884795fcd9fbf1d Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <uwu@icenowy.me>
Date: Tue, 18 Aug 2026 15:27:01 +0800
Subject: [PATCH 1/3] AOSCOS: Revert "aarch64: Fix out-of-range branch veneers
 when --fix-cortex-a53-843419"

This reverts commit d3af4016a645db372ba6812eee651ac5e3eb9159.

With this commit some 843419 sequence will be leaked to the final binary
because of code being moved after being checked.

Signed-off-by: Icenowy Zheng <uwu@icenowy.me>
---
 bfd/elfnn-aarch64.c | 44 +++++++++++++++++++-------------------------
 1 file changed, 19 insertions(+), 25 deletions(-)

diff --git a/bfd/elfnn-aarch64.c b/bfd/elfnn-aarch64.c
index cbd4abea122..2e8ec4d075d 100644
--- a/bfd/elfnn-aarch64.c
+++ b/bfd/elfnn-aarch64.c
@@ -4777,55 +4777,49 @@ elfNN_aarch64_size_stubs (bfd *output_bfd,
       stub_group_size = 127 * 1024 * 1024;
     }
 
-  /* The 843419 erratum fix inserts stub sections in place, not in
-     the section groups.  Running this after we have created the stub
-     groups can perturb the calculations and cause the stub groups
-     that have been created to be out of range.  Avoid this by running
-     this pass first and then creating the groups once we know how much
-     code this mitigation will insert.  */
-  if (htab->fix_erratum_843419 != ERRAT_NONE)
+  group_sections (htab, stub_group_size, stubs_always_before_branch);
+
+  (*htab->layout_sections_again) ();
+
+  if (htab->fix_erratum_835769)
     {
       bfd *input_bfd;
 
       for (input_bfd = info->input_bfds;
-	   input_bfd != NULL;
-	   input_bfd = input_bfd->link.next)
+	   input_bfd != NULL; input_bfd = input_bfd->link.next)
 	{
-	  asection *section;
-
 	  if (!is_aarch64_elf (input_bfd)
 	      || (input_bfd->flags & BFD_LINKER_CREATED) != 0)
 	    continue;
 
-	  for (section = input_bfd->sections;
-	       section != NULL;
-	       section = section->next)
-	    if (!_bfd_aarch64_erratum_843419_scan (input_bfd, section, info))
-	      return false;
+	  if (!_bfd_aarch64_erratum_835769_scan (input_bfd, info,
+						 &num_erratum_835769_fixes))
+	    return false;
 	}
 
       _bfd_aarch64_resize_stubs (htab);
       (*htab->layout_sections_again) ();
     }
 
-  group_sections (htab, stub_group_size, stubs_always_before_branch);
-
-  (*htab->layout_sections_again) ();
-
-  if (htab->fix_erratum_835769)
+  if (htab->fix_erratum_843419 != ERRAT_NONE)
     {
       bfd *input_bfd;
 
       for (input_bfd = info->input_bfds;
-	   input_bfd != NULL; input_bfd = input_bfd->link.next)
+	   input_bfd != NULL;
+	   input_bfd = input_bfd->link.next)
 	{
+	  asection *section;
+
 	  if (!is_aarch64_elf (input_bfd)
 	      || (input_bfd->flags & BFD_LINKER_CREATED) != 0)
 	    continue;
 
-	  if (!_bfd_aarch64_erratum_835769_scan (input_bfd, info,
-						 &num_erratum_835769_fixes))
-	    return false;
+	  for (section = input_bfd->sections;
+	       section != NULL;
+	       section = section->next)
+	    if (!_bfd_aarch64_erratum_843419_scan (input_bfd, section, info))
+	      return false;
 	}
 
       _bfd_aarch64_resize_stubs (htab);
-- 
2.55.0

